File: mobile/lib/models/reports.dart
import '../core/json_parsing.dart';

enum ReportKey { candidates, coordinators, status, biometric, documents }

const List<Map<String, dynamic>> reportDefs = [
  {'key': ReportKey.candidates, 'label': 'Candidates Report'},
  {'key': ReportKey.coordinators, 'label': 'Coordinators Report'},
  {'key': ReportKey.status, 'label': 'Status Distribution Report'},
  {'key': ReportKey.biometric, 'label': 'Biometric Capture Report'},
  {'key': ReportKey.documents, 'label': 'Documents Report'},
];

String reportKeyToPath(ReportKey key) {
  switch (key) {
    case ReportKey.candidates:
      return 'candidates';
    case ReportKey.coordinators:
      return 'coordinators';
    case ReportKey.status:
      return 'status';
    case ReportKey.biometric:
      return 'biometric';
    case ReportKey.documents:
      return 'documents';
  }
}

/// Report rows are rendered generically from whatever columns the backend
/// returns.
typedef ReportRow = Map<String, dynamic>;

class CoordinatorReportRow {
  final int id;
  final String name;
  final String mobile;
  final String? email;
  final String status;
  final int totalCandidates;
  final int completedCandidates;

  CoordinatorReportRow({
    required this.id,
    required this.name,
    required this.mobile,
    this.email,
    required this.status,
    required this.totalCandidates,
    required this.completedCandidates,
  });

  factory CoordinatorReportRow.fromJson(Map<String, dynamic> json) {
    return CoordinatorReportRow(
      id: asInt(json['id']),
      name: json['name'] as String? ?? '',
      mobile: json['mobile'] as String? ?? '',
      email: json['email'] as String?,
      status: json['status'] as String? ?? 'ACTIVE',
      totalCandidates: asInt(json['total_candidates']),
      completedCandidates: asInt(json['completed_candidates']),
    );
  }
}

class MonthlyRegistrationRow {
  final String month;
  final int total;

  MonthlyRegistrationRow({required this.month, required this.total});

  factory MonthlyRegistrationRow.fromJson(Map<String, dynamic> json) {
    return MonthlyRegistrationRow(
      month: json['month']?.toString() ?? '',
      total: asInt(json['total']),
    );
  }
}

class StatusDistributionRow {
  final String status;
  final int total;

  StatusDistributionRow({required this.status, required this.total});

  factory StatusDistributionRow.fromJson(Map<String, dynamic> json) {
    return StatusDistributionRow(
      status: json['status']?.toString() ?? '',
      total: asInt(json['total']),
    );
  }
}

File: frontend/src/types/reports.ts
export type ReportKey = 'candidates' | 'coordinators' | 'status' | 'biometric' | 'documents';

export const REPORT_DEFS: Array<{ key: ReportKey; label: string }> = [
  { key: 'candidates', label: 'Candidates Report' },
  { key: 'coordinators', label: 'Coordinators Report' },
  { key: 'status', label: 'Status Distribution Report' },
  { key: 'biometric', label: 'Biometric Capture Report' },
  { key: 'documents', label: 'Documents Report' },
];

/**
 * Report rows are rendered generically from whatever columns the backend returns,
 * so they are typed as an open record of primitives.
 */
export type ReportRow = Record<string, string | number | boolean | null>;

/** GET /reports/coordinators — also drives the dashboard's coordinator table. */
export interface CoordinatorReportRow {
  id: number;
  name: string;
  mobile: string;
  email: string | null;
  status: 'ACTIVE' | 'INACTIVE';
  total_candidates: number;
  completed_candidates: number;
}

/** GET /reports/monthly-registrations */
export interface MonthlyRegistrationRow {
  month: string;
  total: number;
}

/** Derived client-side from the candidate list for the dashboard pie chart. */
export interface StatusDistributionRow {
  status: string;
  total: number;
}
